[PyTorch][3/4 Static dispatch with multiple backends] Move Tensor method definitions to TensorMethod.cpp

priyaramani opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75607
* __->__ #75606
* #75605

Usecase: Milan models have multiple backends and need to use static dispatch to save on static initialization time and to hit native functions directly from the unboxed APIs.

Cyclic dependency issue: To use multiple backends, TensorMethod definitions (which was previously in TensorBody.h) needed to use multi_dispatch_key_set() API which included List.h. This introduced circular dependency as List.h included Tensor.h down down the line.
Solution: To overcome the circular dependency, this change moves the TensorMethod definitions along with dependency DispatchKeyExtractor.h, from TensorBody.h to TensorMethods.cpp

Differential Revision: [D35512473](https://our.internmc.facebook.com/intern/diff/D35512473/)

priyaramani committed:
[PyTorch][3/4 Static dispatch with multiple backends] Move Tensor method definitions to TensorMethod.cpp

Usecase: Milan models have multiple backends and need to use static dispatch to save on static initialization time and to hit native functions directly from the unboxed APIs.

Cyclic dependency issue: To use multiple backends, TensorMethod definitions (which was previously in TensorBody.h) needed to use multi_dispatch_key_set() API which included List.h. This introduced circular dependency as List.h included Tensor.h down down the line.
Solution: To overcome the circular dependency, this change moves the TensorMethod definitions along with dependency DispatchKeyExtractor.h, from TensorBody.h to TensorMethods.cpp

Differential Revision: [D35512473](https://our.internmc.facebook.com/intern/diff/D35512473/)

[ghstack-poisoned]

priyaramani committed:
Update on "[PyTorch][3/4 Static dispatch with multiple backends] Move Tensor method definitions to TensorMethod.cpp"


Usecase: Milan models have multiple backends and need to use static dispatch to save on static initialization time and to hit native functions directly from the unboxed APIs.

Cyclic dependency issue: To use multiple backends, TensorMethod definitions (which was previously in TensorBody.h) needed to use multi_dispatch_key_set() API which included List.h. This introduced circular dependency as List.h included Tensor.h down down the line.
Solution: To overcome the circular dependency, this change moves the TensorMethod definitions along with dependency DispatchKeyExtractor.h, from TensorBody.h to TensorMethods.cpp

Differential Revision: [D35512473](https://our.internmc.facebook.com/intern/diff/D35512473/)

[ghstack-poisoned]

priyaramani commented:
@priyaramani has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35512473).

bdhirsh commented on tools/codegen/gen.py:
```diff
@@ -1359,15 +1364,12 @@ def static_dispatch_method_headers() -> List[str]:
+            else static_dispatch_extra_headers(static_dispatch_idx, skip_tensor_include=True, skip_dpkey_include=True),
         'tensor_method_declarations': list(mapMaybe(ComputeTensorMethod(
             target=Target.DECLARATION, static_dispatch_backend_indices=static_dispatch_idx), native_functions)),
-        'tensor_method_definitions': list(mapMaybe(ComputeTensorMethod(
```
So after this PR, the call path across static builds vs. dynamic builds will look something like:

dynamic: `at::Tensor::add` (inlined, TensorBody.h) -> `at::_ops::add_Tensor::call` (OperatorEverything.cpp)

static: `at::Tensor::add` (not inlined, TensorBody.h) -> `at::Tensor::add` (definition, TensorMethods.cpp)

We can end up with a much smaller delta between the static vs. non-static builds if we have them both go through the same path. The way you can do that is instead of generating the tensor method definitions differently for static dispatch (making them not inlined, and generating them in `TensorMethods.cpp`), you can:

(1) Update `TensorBody.h` and `Functions.h` to be inlined and call into the `at::_ops` API, exactly the same way it's done in normal core builds

(2) Update the static dispatch codegen to happen inside of the `at::_ops` API, in `Operators.cpp`. Now the static codegen changes are local only to that one file, instead of being spread out. This shouldn't impact perf, since in both cases (your PR vs this version), there's just one function hop.

It shouldn't be too big of a change - you can take most of the existing static dispatch logic, and instead of having it live in `TensorBody.h` and `Functions.h`, you'd have it live around here: https://github.com/pytorch/pytorch/blob/3b18bc36f3effc516b06557073d2f61512afd6d8/tools/codegen/gen.py#L367

(also technically with this PR - on this `tensor_method_definitions` line, you'd have to make them conditionally removed for static dispatch codegen, but keep them in otherwise).

priyaramani replied:
Re having `tensor_method_definitions` conditionally removed for static dispatch codegen : thanks @bdhirsh , will make that change! 

Re having the logic in Operators.cpp, I believe that still won't work like @larryliu0820 mentioned in one of our earlier conversations https://github.com/pytorch/pytorch/pull/73948#discussion_r827584220

larryliu0820 replied:
I guess @bdhirsh has a point. For context, in order to support this model we need the quantized operators and right now they are registered manually (not through codegen). It is non-trivial work to migrate all these quantized ops registration to lightweight dispatch and it's hard to do it clean. Therefore we decided to still leverage the dispatcher for these quantized ops. Previously I thought once we change the logic in `Operators.cpp` we can't use the dispatcher anymore. I've put up more thoughts and it seems to be viable. I do believe `TORCH_LIBRARY` APIs should still work even if we re-generate `Operators.cpp` to static dispatch logic.

bdhirsh replied:
> Previously I thought once we change the logic in Operators.cpp we can't use the dispatcher anymore.

Yep! the stuff in `Operators.cpp` is really just a thin wrapper around calling into the dispatcher (at least, on non-static-dispatch builds).

> Therefore we decided to still leverage the dispatcher for these quantized ops. 

I see- yep, instead of being able to use `at::_ops::{op}::call(...)` to call into the dispatcher on static dispatcher builds, you'd have to use the dispatcher API's directly. it usually looks something like this:
```
  static auto op_handle = c10::Dispatcher::singleton()
      .findSchemaOrThrow(at::_ops::add_Tensor::name, at::_ops::add_Tensor::overload_name)
      .typed<at::_ops::add_Tensor::schema>();
  return op.call(self, other, alpha);
```

Which I agree is a bit wordier than being able to invoke the dispatcher with `at::_ops::`.

I think the benefit of having `at::_ops` use static dispatch though is that there's a much smaller delta between the two builds, which leaves less room for bugs and also makes system easier to reason about w.r.t. any future changes. It sounds like you think the "using the dispatcher for quantized ops" is some short-to-mid term temporary state for mobile?

larryliu0820 replied:
> It sounds like you think the "using the dispatcher for quantized ops" is some short-to-mid term temporary state for mobile?

I think it's a mid term state, we need to properly design a macro that handles these non-codegen case properly, if we want to use lightweight dispatch for these quantized operators.

But yeah generally agree that we don't have a blocker to migrate all the static dispatch logic into `Operators.cpp`.

priyaramani committed:
Update on "[PyTorch][3/4 Static dispatch with multiple backends] Move Tensor method definitions to TensorMethod.cpp"


Usecase: Milan models have multiple backends and need to use static dispatch to save on static initialization time and to hit native functions directly from the unboxed APIs.

Cyclic dependency issue: To use multiple backends, TensorMethod definitions (which was previously in TensorBody.h) needed to use multi_dispatch_key_set() API which included List.h. This introduced circular dependency as List.h included Tensor.h down down the line.
Solution: To overcome the circular dependency, this change moves the TensorMethod definitions along with dependency DispatchKeyExtractor.h, from TensorBody.h to TensorMethods.cpp

Differential Revision: [D35512473](https://our.internmc.facebook.com/intern/diff/D35512473/)

[ghstack-poisoned]

priyaramani committed:
Update on "[PyTorch][3/4 Static dispatch with multiple backends] Move Tensor method definitions to TensorMethod.cpp"


Usecase: Milan models have multiple backends and need to use static dispatch to save on static initialization time and to hit native functions directly from the unboxed APIs.

Cyclic dependency issue: To use multiple backends, TensorMethod definitions (which was previously in TensorBody.h) needed to use multi_dispatch_key_set() API which included List.h. This introduced circular dependency as List.h included Tensor.h down down the line.
Solution: To overcome the circular dependency, this change moves the TensorMethod definitions along with dependency DispatchKeyExtractor.h, from TensorBody.h to TensorMethods.cpp

Differential Revision: [D35512473](https://our.internmc.facebook.com/intern/diff/D35512473/)

[ghstack-poisoned]

priyaramani committed:
Update on "[PyTorch][3/4 Static dispatch with multiple backends] Move Tensor method definitions to TensorMethod.cpp"


Usecase: Milan models have multiple backends and need to use static dispatch to save on static initialization time and to hit native functions directly from the unboxed APIs.

Cyclic dependency issue: To use multiple backends, TensorMethod definitions (which was previously in TensorBody.h) needed to use multi_dispatch_key_set() API which included List.h. This introduced circular dependency as List.h included Tensor.h down down the line.
Solution: To overcome the circular dependency, this change moves the TensorMethod definitions along with dependency DispatchKeyExtractor.h, from TensorBody.h to TensorMethods.cpp

Differential Revision: [D35512473](https://our.internmc.facebook.com/intern/diff/D35512473/)

[ghstack-poisoned]
